File: 1249-minimum-remove-to-make-valid-parentheses/1249-minimum-remove-to-make-valid-parentheses.js
/**
 * @param {string} s
 * @return {string}
 */
var minRemoveToMakeValid = function(s) {
    const stack = [];
    const chars = [...s];

    for(let i = 0; i<chars.length; i++) {
        if(chars[i] === '(') stack.push(i);
        else if(chars[i] === ')') {
            if(stack.length > 0) stack.pop();
            else chars[i] = '*'
        }
    }

    while(stack.length > 0) chars[stack.pop()] = '*';

    let ans = "";
    for(let i = 0; i<chars.length; i++) {
       if(chars[i] !== '*') ans += chars[i];
    }

    return ans;
};

File: 1249-minimum-remove-to-make-valid-parentheses/1249-minimum-remove-to-make-valid-parentheses.java
class Solution(object):
    def minRemoveToMakeValid(self, s):
        stack = []
        chars = list(s)

        for i in range(len(chars)):
            if chars[i] == '(':
                stack.append(i)
            elif chars[i] == ')':
                if stack:
                    stack.pop()
                else:
                    chars[i] = '*'

        while stack:
            chars[stack.pop()] = '*'

        return ''.join(char for char in chars if char != '*')

        